react/week4/mainForm.js

```diff
@@ -0,0 +1,28 @@
+import React, { useContext } from "react";
+import { StateContext } from "./context.js";
+
+export default function MainForm() {
+  const { users, loading, fetchUsers } = useContext(StateContext);
+
+  return (
+    <div className="App">
+      <h1>Github user searcher</h1>
+      <div>
+        <input
+          placeholder="Search for user"
+          onKeyUp={(e) => {
+            fetchUsers(e.target.value);
+          }}
+        />
+      </div>
+      <div>
+        {loading && <div>Loading... </div>}
+        {users.length === 0 || !users ? (
+          <h2>No results</h2>
+        ) : (
+          users.map((user) => <div key={user.id}> {user.login}</div>)
```

Comment:
See if you can add some debouncing logic as well. You can reuse the logic we saw with the timer with the timeout being cleared in the cleanup function.

Reply:
Thanks for your time and consideration. I guess, I didn't understand the concept of the debouncing logic well. May I get a bit more of a help from you about it?

Reply:
The above with debouncing. This one is more advanced.

```
import React, { useContext, createContext, useState, useEffect } from "react";

function App() {
  return (
    <UsersSearch/>
  )
}

export const UserSearchContext = createContext();

function UsersSearch() {
  const [userInput, setUserInput] = useState("");
  const [users, isCalling, isDebouncing] = useDebouncer(userInput, request);

  const handleErrors = (response) => {
    if (!response.ok) {
      throw Error(response.statusText);
    }
    return response;
  };


    async function request () {
      try {
        const response = await fetch(
          `https://api.github.com/search/users?q=${userInput}`
        );
        handleErrors(response);
        const data = await response.json();
        return await data.items;
      } catch (err) {
        console.log(err);
      } finally {
      }
    }

 
  return (
    <UserSearchContext.Provider value={{ users, userInput, setUserInput }}>
      <SearchForm />
      { isCalling || isDebouncing ? <Loading /> : <UsersDisplay /> }
    </UserSearchContext.Provider>
  );
}

function Loading() {
  return (
    <div>Loading... </div>
  )
}

function UsersDisplay() {
    const { users} = useContext(UserSearchContext);
  
    return (
      <div className="App">
        {!users  || users.length === 0 ? (
          <h2>No results</h2>
        ) : (
          users.map((user) => <div key={user.id}> {user.login}</div>)
        )}
      </div>
    );
  }

function SearchForm() {
    const { userInput, setUserInput } = useContext(UserSearchContext);
  
    return (
      <div className="App">
        <h1>Github user searcher</h1>
        <div>
          <input
            placeholder="Search for user"
            value={userInput}
            onChange={(e) => {
              setUserInput(e.target.value);
            }}
          />
        </div>
      </div>
    );
  }

  function useDebouncer(inputText, request) {
    const [isCalling, setIsCalling] = useState(false);
    const [callQuery, setCallQuery] = useState('');
    const [response, setResponse] = useState('');
    const [isDebouncing, setIsDebouncing] = useState(false)

    useEffect(() => {
      if (isCalling) return;
      if (callQuery === inputText) return;
      setIsDebouncing(true);
      const timerId = setTimeout(async () => {
          setIsDebouncing(false);
          setIsCalling(true);
          setCallQuery(inputText);
          const response = inputText !== "" ? await request() : "";
          setResponse(response);
          setIsCalling(false);
      }, 500);
      return () => clearTimeout(timerId);
    }, [inputText, isCalling]);
    
    return [response, isCalling, isDebouncing];
  }

export default App;
```

Reply:
Thanks for your great support. I'm still missing some knowledge about debouncer.